src/gui/windows/predictor/predictor_btb_dock.cpp

```diff
@@ -6,158 +6,158 @@ DockPredictorBTB::DockPredictorBTB(QWidget *parent) : Super(parent) {
-    machine::Address target_address) {
-    if (index >= btb->rowCount()) {
-        WARN("BTB dock update received invalid row index: %u", index);
+}
+
+void DockPredictorBTB::update_btb_row(
+    uint16_t row_index,
+    machine::BranchTargetBufferEntry btb_entry
+) {
+    if (row_index >= btb->rowCount()) {
+        WARN("BTB dock update received invalid row index: %u", row_index);
         return;
     }
 
-    set_row_color(index, Q_COLOR_UPDATE);
+    QTableWidgetItem *item;
 
-    QTableWidgetItem *item_index { btb->item(index, DOCK_BTB_COL_INDEX) };
-    QTableWidgetItem *item_instr_addr { btb->item(index, DOCK_BTB_COL_INSTR_ADDR) };
-    QTableWidgetItem *item_target_addr { btb->item(index, DOCK_BTB_COL_TARGET_ADDR) };
+    if (btb_entry.entry_valid) {
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_INSTR_ADDR);
+        item->setData(Qt::DisplayRole, machine::addr_to_hex_str(btb_entry.instruction_address));
 
-    if (item_index == nullptr) {
-        item_index = new QTableWidgetItem();
-        btb->setItem(index, DOCK_BTB_COL_INDEX, item_index);
-    }
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_TARGET_ADDR);
+        item->setData(Qt::DisplayRole, machine::addr_to_hex_str(btb_entry.target_address));
 
-    if (item_instr_addr == nullptr) {
-        item_instr_addr = new QTableWidgetItem();
-        btb->setItem(index, DOCK_BTB_COL_INSTR_ADDR, item_instr_addr);
-    }
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_TYPE);
+        item->setData(Qt::DisplayRole, machine::branch_type_to_string(btb_entry.branch_type).toString());
+    } else {
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_INSTR_ADDR);
+        item->setData(Qt::DisplayRole, "");
 
-    if (item_target_addr == nullptr) {
-        item_target_addr = new QTableWidgetItem();
-        btb->setItem(index, DOCK_BTB_COL_TARGET_ADDR, item_target_addr);
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_TARGET_ADDR);
+        item->setData(Qt::DisplayRole, "");
+
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_TYPE);
+        item->setData(Qt::DisplayRole, "");
     }
+}
 
-    item_instr_addr->setData(Qt::DisplayRole, machine::addr_to_hex_str(instruction_address));
-    item_target_addr->setData(Qt::DisplayRole, machine::addr_to_hex_str(target_address));
-};
+void DockPredictorBTB::highligh_row_after_prediction(uint16_t row_index) {
+    set_row_color(row_index, Q_COLOR_PREDICT);
+}
 
-void DockPredictorBTB::highligh_row_after_prediction(uint16_t index) {
-    set_row_color(index, Q_COLOR_PREDICT);
+void DockPredictorBTB::highligh_row_after_update(uint16_t row_index) {
+    set_row_color(row_index, Q_COLOR_UPDATE);
 }
 
 void DockPredictorBTB::reset_colors() {
     set_table_color(Q_COLOR_DEFAULT);
+}
+
+void DockPredictorBTB::clear_btb() {
+    for (uint16_t row_index = 0; row_index < btb->rowCount(); row_index++) {
+        QTableWidgetItem *item;
+
+        item = get_btb_cell_item(row_index, DOCK_BTB_COL_INDEX);
```

Comment:
Same as above. Btw, given this use, it might be good to make a function that takes 4 strings and sets the all of them. Not blocking.